### PR TITLE
build: add `tests` extras & update `developer_docs.md`

### PR DESCRIPTION
**Pull request by alvarobartt**
# Description

Since the only option to setup a development environment was via `conda` and not all the users don't use `conda`, I've included the extra dependencies to run the tests as extras, to be able to easily install those via `pip` without forcing the users to rely on `conda`.

Besides that, I've also included some more information in the `developer_docs.md` to make it easier for the users to install everything and run the tests locally.

Additionally, while tackling this PR, some other minor things have been spotted and solved in other PRs:

* https://github.com/argilla-io/argilla/pull/3015
* https://github.com/argilla-io/argilla/pull/3012
* https://github.com/argilla-io/argilla/pull/3105

**Type of change**

- [X] Improvement (change adding some improvement to an existing functionality)
- [X] Documentation update

**How Has This Been Tested**

- [x] Tested in Python 3.8, 3.9 and 3.10 by installing everything as `pip install -e ".[server,postgresql,listeners,tests]"` and then running `pytest`
- [X] Tested that `conda` installation still works fine

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -6,8 +6,8 @@ build-backend = "setuptools.build_meta"
 name = "argilla"
 description = "Open-source tool for exploring, labeling, and monitoring data for NLP projects."
 readme = "README.md"
-requires-python = ">=3.7"
-license = {text = "Apache-2.0"}
+requires-python = ">=3.7"  # TODO: upgrade minimum Python version to 3.8, and pin maximum to < 3.12
```
Comment (gabrielmbmb) [inline pyproject.toml:9]: We will deprecate `3.7` in the `1.9.0` release, right?

**Review by gabrielmbmb — APPROVED** [review]
Hey @alvarobartt, LGTM! This will be really helpful 😄 Maybe it would be good also to mention in the developer docs something about the dependencies needed to be installed to be able to build and install `psycopg2`.

File: pyproject.toml (hunk 1 above)
Comment (alvarobartt) [reply pyproject.toml:9]: Yes I think so, didn't want to update that yet!

**Comment by alvarobartt** [comment]
> Hey @alvarobartt, LGTM! This will be really helpful 😄 Maybe it would be good also to mention in the developer docs something about the dependencies needed to be installed to be able to build and install `psycopg2`.

Sure, thanks for the feedback @gabrielmbmb! I'll push a couple more commits and ask you to re-review 😄

**Review by gabrielmbmb — APPROVED** [review]

**Review by gabrielmbmb — APPROVED** [review]

**Comment by alvarobartt** [comment]
cc @gabrielmbmb this works fine up until Python 3.10, I've tested in 3.11 and it doesn't work well, I guess it's related to `psycopg-binary` and some required dependencies in macOS, can you please check? Thanks

**Pull request merged**